Let’s make it possible to read a file using the stdlib

gasche opened the pull request:
In Stdlib:

```ocaml
val really_input_up_to : in_channel -> bytes -> int -> int -> int
(** [really_input ic buf pos len] reads at most [len] characters
   from the given channel [ic], storing them in byte sequence [buf], starting at
   character number [pos].
   Compared to [input], this function only stops when no more characters
   are available -- the end of file is reached.
   The return value indicates the number of characters that were available.
   The exception [Invalid_argument "really_input_up_to"] is raised
   if [pos] and [len] do not designate a valid range of [buf].

   @since 4.09
*)
```

In Buffer:

```ocaml
val add_channel_full : t -> in_channel -> int
(** [add_channel b ic] reads all the available characters from
   the input channel [ic] and stores them at the end of buffer [b].
   Returns the number of characters read.

   @since 4.09
*)
```

gasche committed:
stdlib: input_up_to, input_string_up_to

  val input_up_to : in_channel -> bytes -> pos:int -> len:int -> int
  val input_string_up_to : in_channel -> int -> string

Consumes all the characters available on the input channel.

gasche committed:
use input_up_to in ocamltest/filecompare.ml

trefis commented on stdlib/stdlib.mli:
```diff
@@ -1060,16 +1060,29 @@ val input : in_channel -> bytes -> int -> int -> int
    if desired.  (See also {!Stdlib.really_input} for reading
    exactly [len] characters.)
-   Exception [Invalid_argument "input"] is raised if [pos] and [len]
+   The exception [Invalid_argument "input"] is raised if [pos] and [len]
```
Why not use `@raise`?

gasche replied:
I just reused the existing documentation comments. I may change the three I touched (input, really_input, really_input_up_to) to use `@raise`.

gasche replied:
Done.

hannesm commented:
I'd appreciate if the `offset` and `length` arguments would be labeled - maybe as well making them optional and provide sane default values

dra27 commented on stdlib/buffer.mli:
```diff
@@ -148,6 +148,14 @@ val add_channel : t -> in_channel -> int -> unit
    the buffer, so as to avoid loss of data. *)
 
+val add_channel_full : t -> in_channel -> int
+(** [add_channel b ic] reads all the available characters from
```
This should be `add_channel_full`. I would be more terse and define it with reference to `add_channel` - e.g. `[add_channel_full] is as [add_channel] but returns the number of characters read.`

gasche replied:
The terser explanation you give is not really correct, as `add_channel_full` does not take a limit on the total length to read. I fixed the copy-error.

dra27 replied:
Sorry - I realised this morning. It turns out reviewing on one's phone limits the visible text...

dra27 commented on stdlib/stdlib.mli:
```diff
@@ -1060,16 +1060,29 @@ val input : in_channel -> bytes -> int -> int -> int
    do not designate a valid range of [buf]. *)
 
+val really_input_up_to : in_channel -> bytes -> int -> int -> int
+(** [really_input ic buf pos len] reads at most [len] characters
```
This should be `really_input_up_to`

dra27 reviewed (commented):
I don’t like either of the names, but I’m struggling with better suggestions at the moment. `input_max`/`input_best`/`input_greedy` - something without the word really, given that it might not really happen? `add_channel_full` sounds it like it does more than `add_channel` but `add_channeli` is my best awful alternate so far!!

Drup commented:
What about `input_into : ?pos:int -> ?len:int -> channel -> bytes -> int` and we deprecate `input` and `really_input` ? Do we really need 3 low-level input functions ?

gasche commented:
Regarding Stdlib input functions, several things can happen when trying to read up to `len` bytes:

- Early end: the end of input is reached before `len` bytes could be read.
- Partial read: the system decides that returning a (non-empty) partial read is better (for performance reasons?).

Currently the standard library has `input`, which returns control to the user in both cases of early ends and partial reads, and `really_input`, which handles "partial reads" itself by looping, and fails with an exception in the "early end" case -- and does not provide a way to know the amount of data read, so catching the exception does not make it usable.

The idea of the new function is to handle partial reads (like `really_input`), but return control to the user in the case of early end (like `input` and unlike `really_input`). This is useful in all cases where the user wishes to bound the amount of data read (it goes in a fixed-sized buffer somewhere), but does not have a better way to handle partial reads than reading more. I initially named this function `really_input_up_to`, and just renamed it into `input_up_to` in the vain hope of pleasing our noble bikeshedding group :-)

In the compiler distribution, there is a use-case in ocamltest/file_compare, which implemented its own variant of `input_up_to` (see the patch for details), and in the new `Buffer.add_buffer_full` function that I provide, which cannot be implemented with `really_input` but is painful to implement with `input` (and actually this creates potential re-entrency bugs, which is what initiated this PR) .

For good measure I added a function `input_string_up_to`, which is to `input_up_to` what `really_input_string` is to `really_input` (sometimes it has to call the equivalent of `String.sub`, so it may allocate two strings instead of one; note that this only happens when the end of input is reached, so not very often in looping scenarios).

@hannesm: I did change to named arguments `input_channel -> bytes -> pos:int -> len:int -> int`, but I'm not fond of the result (it is inconsistent with other functions and the names are not great; in particular `len` is only the maximal limit on the read length, not the actual length) so I may revert back eventually.

@Drup: yes, I believe that having three functions is reasonable. `input` is the OCaml equivalent of C input functions, and it is the most flexible of all three, so it should stay. `really_input{,_string}` is used often in the compiler distribution, typically to read magic numbers. The new functions have their use-cases, and it is more flexible than `really_input`, but it does not clearly subsume either of them.

(I'm not sure why people are so excited about deprecating stuff, but I'm not, so I would rather leave deprecation off the table for this PR.)

Now on `Buffer.add_channel_full`: it is not possible in Stdlib to add a function that consumes "all" an input channel, as long as we only read into fixed-size byte containers. This is the job of the dynamic containers of the `Buffer` module. This is exactly what `add_channel_full` provides -- in addition to the existing `add_channel : Buffer.t -> input_channel -> int -> unit`, which is length-limited. Note that the implementation internally uses `input_up_to` -- it is incorrect to use `really_input` there, see #6719.

dra27 commented on stdlib/buffer.ml:
```diff
@@ -197,6 +197,15 @@ let add_channel b ic len =
   add_channel_rec b ic len
 
+let add_channel_full b ic =
+  let rec loop b ic already_read =
```
Why pass `b` in `loop`, given that it's a mutable structure?

dra27 commented on stdlib/buffer.ml:
```diff
@@ -197,6 +197,15 @@ let add_channel b ic len =
+    let already_read = already_read + n in
+    if n < len then n
+    else loop b ic already_read
+  in loop b ic 0
```
The `80` caught my eye, but something else cannot be right - there's no call to `resize`?

I got lost in the various issues leading to this PR - what's wrong with using `input` and just accepting whatever it gives back for the loop? It seems strange to be using one function `input_up_to` which loops over `input` to get some input and then putting another loop around that to keep reading?

gasche replied:
You are very correct, various things in this implementation were wrong (`then n` was also wrong). Apologies for the confusion; it looks like I've gone back-and-forth between a bit too many different choices to think clearly about it yesterday evening.

I have done another iteration that should fix the implementation. I'm also planning to write tests to exercise these functions, but I was sort of hoping that the discussion on "do we want these functions?" and "what names should they have?" would sort of settle down before we start tearing the implementations apart.

> what's wrong with using `input` and just accepting whatever it gives back for the loop

My initial idea was to move the "loop on partial input" logic from the Buffer module to the Stdlib module (and have other potential consumers benefit from it). As you suggest, we have to loop anyway so it may be just as easy to use `input`. I'll give it a try.

cfcs replied:
@gasche Ooh, a general `loop on partial input` utility function in the stdlib sounds like an interesting idea.

dra27 reviewed (commented):
Something doesn't seem right in `Buffer.add_channel_full` - sorry if I've missed something.

gasche commented:
Thanks to @dra27's sanity-checking I changed `Buffer.add_channel_full` to use `input` instead of `input_up_to`. This means that this function is independent from the `Stdlib` changes (indeed it started that way) so I could split the PR in two.

dra27 commented:
Buffer change looks good - up to you if you want to split it off for quick merging while a naming/API decision settles for the rest. (my previous objection to the name of `Buffer.add_channel_full` was itself flawed)

gasche commented:
I wouldn't pass on the chance of making *two* stdlib PRs in a single week. My biggest problem is that I already used the most awesome branch name for this one PR; I feel the name should go with the `Buffer` change but we mostly discussed `input` functions here...

gasche commented:
I moved the `Buffer.add_channel_full` function to #8600, so that this PR can focus on `input_up_to` and `input_string_up_to`. Thanks everyone for the feedback so far!

hannesm commented:
@gasche maybe `offset:int` and `max_bytes:int` are better labels? There's some struggle with already existing Stdlib bindings that _don't_ use labels (yet ;)), and thus it feels inconsistent. IMHO it is worth to add the labels to avoid further confusion (esp. when reading code, I always have to look up the documentation about the ordering of arguments in bindings of the Stdlib.

dbuenzli commented:
I don't think this should be designed in isolation of https://github.com/ocaml/ocaml/pull/640. More specifically I'd prefer if the `Channel` module scheme [mentioned](https://github.com/ocaml/ocaml/pull/640#issuecomment-433601405) there would first be implemented (it seems to have @alainfrisch's agreement) so that we don't add more things to "pervasives".

gasche commented:
@dbuenzli's point is reasonable, so I'm closing this for the time being.
